core/trino-main/src/main/java/io/trino/execution/SqlTaskExecution.java

```diff
@@ -643,6 +602,31 @@ public Driver createDriver(DriverContext driverContext, @Nullable ScheduledSplit
             return driver;
         }
 
+        public void addSplits(Set<ScheduledSplit> splits, boolean noMoreSplits)
+        {
+            List<Driver> drivers = new ArrayList<>();
+            synchronized (this) {
+                for (WeakReference<Driver> driverReference : createdDrivers) {
+                    Driver driver = driverReference.get();
+                    if (driver != null) {
+                        drivers.add(driver);
+                    }
+                }
+                if (drivers.isEmpty()) {
+                    pendingSplits.addAll(splits);
+                }
+                this.noMoreSplits = noMoreSplits;
+            }
+            boolean splitsAdded = false;
+            for (Driver driver : drivers) {
+                driver.updateSplitAssignment(new SplitAssignment(
+                        driverFactory.getSourceId().orElseThrow(),
+                        splitsAdded ? ImmutableSet.of() : splits,
```

Comment:
why assigning all splits to single driver?

Reply:
Changed